core/src/docs/rfcs/3197_config.md

```diff
@@ -0,0 +1,234 @@
+    Azblob(StorageAzblobConfig),
+    Fs(StorageFsConfig),
+    Ftp(StorageFtpConfig),
+    Gcs(StorageGcsConfig),
+    Hdfs(StorageHdfsConfig),
+    Http(StorageHttpConfig),
+    Ipfs(StorageIpfsConfig),
+    Memory,
+    Moka(StorageMokaConfig),
+    Obs(StorageObsConfig),
+    Oss(StorageOssConfig),
+    S3(StorageS3Config),
+    Redis(StorageRedisConfig),
+    Webhdfs(StorageWebhdfsConfig),
+    Cos(StorageCosConfig),
+}
+```
+
+So I propose to expose services config to the users, allowing them to work on config structs directly and fetch the config at runtime.
+
+# Guide-level explanation
+
+First of all, we will add config struct for each service. For example, `Fs` will have a `FsConfig` struct and `S3` will have a `S3Config`. The fields within the config struct are public and marked as non-exhaustive.
+
+```rust
+#[non_exhaustive]
+pub struct S3Config {
+  pub root: Option<String>,
+  pub bucket: String,
+  pub endpoint: Option<String>,
+  pub region: Option<String>,
+  ...
+}
+```
+
+Then, we will add a `Config` enum that contains all the config structs. The enum is public and non-exhaustive too.
+
+```rust
+#[non_exhaustive]
+pub enum Config {
+  Fs(FsConfig)
+  S3(S3Config),
+  Custom(&'static str, HashMap<String, String>),
+}
+```
+
+Notably, a `Custom` variant will be added to the enum. This variant aligns with `Scheme::Custom(name)` and allows users to configure custom services.
+
+At `Operator` level, we will add `from_config` and `via_config` methods.
+
+```rust
+impl Operator {
+  pub fn via_config(cfg: impl Into<Config>) -> Result<Operator> {}
+}
+```
+
+Additionally, `OperatorInfo` will introduce a new API method, `config()`:
+
+```rust
+impl OperatorInfo {
+  pub fn config(&self) -> Config {}
+}
+```
+
+Users can use `config()` to fetch the config of a service at runtime and construct a new operator based on needs.
+
+# Reference-level explanation
+
+Every services will have a `XxxConfig` struct.
+
+`XxxConfig` will implement the following things:
+
+- `Default` trait: All config fields will have a default value.
+- `Serialize` trait.
+- `Deserialize` trait.
+- `FromMap` trait: Convert from a `HashMap<String, String>`.
+- `Into<Config>` trait: All service config can be converted to `Config` enum.
+
+All config fields will be public and non-exhaustive, allowing users to build config this way:
+
+```rust
+let s3 = S3Config {
+  bucket: "test".to_string(),
+  endpoint: Some("http://localhost:9000".to_string()),
+  ..Default::default()
+}
+```
+
+The public API of existing builders will remain unchanged, although their internal implementations will be modified to utilize `XxxConfig`. Type that can't be represents as `String` like `Box<dyn AwsCredentialLoad>` and `HttpClient` will be kept in `Builder` as before.
+
+For example:
+
+```rust
+#[non_exhaustive]
+pub struct S3Config {
+  pub root: Option<String>,
+  pub bucket: String,
+  pub endpoint: Option<String>,
+  pub region: Option<String>,
+  ...
+}
+
+pub struct S3Builder {
```

Comment:
Mostly LGTM, I still think this config semantic is too complicated for the user. I think the hashmap config function is more straightforward for the user， Maybe we can implement a schema validation for the hashmap config to protect OpenDAL config loading security (Although this is not type safe, it can still ensure the input safety of the system.)

Reply:
> Maybe we can implement a schema validation for the hashmap config to protect OpenDAL config loading security (Although this is not type safe, it can still ensure the input safety of the system.)

While it is possible and a good idea to add schema validation for the `map` construct method, OpenDAL still need to offer a new configuration approach that addresses different problems.

Consider this scenario: OpenDAL implemented a breaking change by renaming the configuration from `abc` to `def`. Users' applications will continue to compile without issues until their end users discover that `abc` does not function as intended.

This is really hard to discover. NOTE: Our users may have hundreds of dependencies, making it impossible to revisit every breaking change listed in OpenDAL. To make matters worse, OpenDAL is being increasingly utilized in other libraries. This implies that if a library called `closedal` relies on `opendal` and updates its version, all applications dependent on `closedal` will also be impacted.

So we need to provide our users a new method that:

- They can discover which config that OpenDAL supports. 
- They can know at compile time about the breaking changes for config.

The `Builder` pattern meets those requirements, but its code is provided in a way that leads to duplication for all our users. `Config` makes is possible for users to define their our config struct and implement `TryFrom/TryInto` instead.